src/Functions/sparseGrams.cpp

```diff
@@ -0,0 +1,271 @@
+#include <optional>
+#include <DataTypes/DataTypesNumber.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/FunctionTokens.h>
+#include <Common/Exception.h>
+
+#include <Poco/UTF8Encoding.h>
+#include <zlib.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int BAD_ARGUMENTS;
+extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+/** Functions that finds all substrings win minimal length n
+  * such their border (n-1)-grams' hashes are more than hashes of every (n-1)-grams' in substring.
+  * As a hash function use zlib crc32, which is crc32-ieee with 0xffffffff as initial value
+  *
+  * sparseGrams(s)
+  */
+namespace
+{
+
+using Pos = const char *;
+
+template <bool is_utf8>
+class SparseGramsImpl
+{
+private:
+    Pos pos;
+    Pos end;
+    std::vector<UInt32> ngram_hashes;
```

Comment:
Thinking about the problem, it can be optimized (there is no need to do it here). For example, in corner cases that produce a lot of repetitive n-grams, a hashmap can help.

Reply:
I think it doesn't make sense to store hashes in a hashmap. It means you have to calculate hash to get hash from the map :)
crc32 is a fast algo, maybe we don't need to store it at all, just left and right to avoid double calculation